File: app/src/main/java/me/ykrank/s1next/data/db/dbmodel/BlackWord.java
package me.ykrank.s1next.data.db.dbmodel;

import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.IntDef;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;

import org.greenrobot.greendao.annotation.Entity;
import org.greenrobot.greendao.annotation.Generated;
import org.greenrobot.greendao.annotation.Id;
import org.greenrobot.greendao.annotation.Index;
import org.greenrobot.greendao.annotation.Property;
import org.jetbrains.annotations.NotNull;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

import me.ykrank.s1next.R;

@Entity(nameInDb = "BlackWord")
public class BlackWord implements Parcelable {

    public static final int NORMAL = 0;
    public static final int HIDE = 1;
    public static final int DEL = 2;

    private static final String TimeFormat = "yyyy-MM-dd HH:mm";

    @Id(autoincrement = true)
    @Nullable
    private Long id;

    /**
     * 屏蔽词
     */
    @Property(nameInDb = "Word")
    @Index(name = "IDX_BlackWord_Word", unique = true)
    private String word;

    /**
     * 屏蔽状态
     */
    @Property(nameInDb = "Stat")
    @BlackWordFLag
    private int stat = NORMAL;

    /**
     * 屏蔽时的时间
     */
    @Property(nameInDb = "Timestamp")
    private long timestamp;

    /**
     * 是否已同步
     */
    @Property(nameInDb = "Upload")
    private boolean upload;

    public BlackWord() {
        this.timestamp = System.currentTimeMillis();
    }

    public BlackWord(String word, int stat) {
        this.word = word;
        this.stat = stat;
        this.timestamp = System.currentTimeMillis();
        this.upload = false;
    }

    protected BlackWord(Parcel in) {
        if (in.readByte() == 0) {
            id = null;
        } else {
            id = in.readLong();
        }
        word = in.readString();
        stat = in.readInt();
        timestamp = in.readLong();
        upload = in.readByte() != 0;
    }

    @Generated(hash = 1127251493)
    public BlackWord(Long id, String word, int stat, long timestamp, boolean upload) {
        this.id = id;
        this.word = word;
        this.stat = stat;
        this.timestamp = timestamp;
        this.upload = upload;
    }

    public static final Creator<BlackWord> CREATOR = new Creator<BlackWord>() {
        @Override
        public BlackWord createFromParcel(Parcel in) {
            return new BlackWord(in);
        }

        @Override
        public BlackWord[] newArray(int size) {
            return new BlackWord[size];
        }
    };

    @Nullable
    public Long getId() {
        return id;
    }

    public void setId(@Nullable Long id) {
        this.id = id;
    }

    public String getWord() {
        return word;
    }

    public void setWord(String word) {
        this.word = word;
    }

    public int getStat() {
        return stat;
    }

    public void setStat(int stat) {
        this.stat = stat;
    }

    @StringRes
    public int getStatRes() {
        switch (stat) {
            case HIDE:
                return R.string.blacklist_flag_hide;
            case DEL:
                return R.string.blacklist_flag_del;
            default:
                return R.string.blacklist_flag_normal;
        }
    }

    public long getTimestamp() {
        return timestamp;
    }

    public void setTimestamp(long timestamp) {
        this.timestamp = timestamp;
    }

    public String getTime() {
        SimpleDateFormat sdf = new SimpleDateFormat(TimeFormat, Locale.getDefault());
        return sdf.format(new Date(timestamp));
    }

    public boolean isUpload() {
        return upload;
    }

    public void setUpload(boolean upload) {
        this.upload = upload;
    }

    @Override
    public String toString() {
        return "BlackWord{" +
                "id=" + id +
                ", word='" + word + '\'' +
                ", stat=" + stat +
                ", timestamp=" + timestamp +
                ", upload=" + upload +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        BlackWord blackWord = (BlackWord) o;

        if (stat != blackWord.stat) return false;
        if (timestamp != blackWord.timestamp) return false;
        if (upload != blackWord.upload) return false;
        if (id != null ? !id.equals(blackWord.id) : blackWord.id != null) return false;
        return word != null ? word.equals(blackWord.word) : blackWord.word == null;
    }

    @Override
    public int hashCode() {
        int result = id != null ? id.hashCode() : 0;
        result = 31 * result + (word != null ? word.hashCode() : 0);
        result = 31 * result + stat;
        result = 31 * result + (int) (timestamp ^ (timestamp >>> 32));
        result = 31 * result + (upload ? 1 : 0);
        return result;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        if (id == null) {
            dest.writeByte((byte) 0);
        } else {
            dest.writeByte((byte) 1);
            dest.writeLong(id);
        }
        dest.writeString(word);
        dest.writeInt(stat);
        dest.writeLong(timestamp);
        dest.writeByte((byte) (upload ? 1 : 0));
    }

    public void mergeFrom(@NotNull BlackWord blackWord) {
        this.word = blackWord.word;
        this.stat = blackWord.stat;
        this.timestamp = blackWord.timestamp;
    }

    public boolean getUpload() {
        return this.upload;
    }

    @IntDef({NORMAL, HIDE, DEL})
    public @interface BlackWordFLag {

    }
}

File: app/src/main/java/me/ykrank/s1next/data/db/dbmodel/BlackWord.kt
package me.ykrank.s1next.data.db.dbmodel

import android.os.Parcel
import android.os.Parcelable
import androidx.annotation.IntDef
import androidx.annotation.StringRes
import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Index
import androidx.room.PrimaryKey
import me.ykrank.s1next.R
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

@Entity(
    tableName = "BlackWord",
    indices = [
        Index(value = ["Word"], name = "IDX_BlackWord_Word", unique = true),
    ]
)
class BlackWord : Parcelable {
    @PrimaryKey(autoGenerate = true)
    @ColumnInfo(name = "_id")
    var id: Long? = null

    /**
     * 屏蔽词
     */
    @ColumnInfo(name = "Word")
    var word: String? = null

    /**
     * 屏蔽状态
     */
    @ColumnInfo(name = "Stat")
    @BlackWordFLag
    var stat = NORMAL

    /**
     * 屏蔽时的时间
     */
    @ColumnInfo(name = "Timestamp")
    var timestamp: Long

    /**
     * 是否已同步
     */
    @ColumnInfo(name = "Upload")
    var isUpload = false

    constructor() {
        timestamp = System.currentTimeMillis()
    }

    constructor(word: String?, stat: Int) {
        this.word = word
        this.stat = stat
        timestamp = System.currentTimeMillis()
        isUpload = false
    }

    protected constructor(`in`: Parcel) {
        id = if (`in`.readByte().toInt() == 0) {
            null
        } else {
            `in`.readLong()
        }
        word = `in`.readString()
        stat = `in`.readInt()
        timestamp = `in`.readLong()
        isUpload = `in`.readByte().toInt() != 0
    }

    constructor(id: Long?, word: String?, stat: Int, timestamp: Long, upload: Boolean) {
        this.id = id
        this.word = word
        this.stat = stat
        this.timestamp = timestamp
        isUpload = upload
    }

    @get:StringRes
    val statRes: Int
        get() = when (stat) {
            HIDE -> R.string.blacklist_flag_hide
            DEL -> R.string.blacklist_flag_del
            else -> R.string.blacklist_flag_normal
        }
    val time: String
        get() {
            val sdf = SimpleDateFormat(TimeFormat, Locale.getDefault())
            return sdf.format(Date(timestamp))
        }

    override fun toString(): String {
        return "BlackWord{" +
                "id=" + id +
                ", word='" + word + '\'' +
                ", stat=" + stat +
                ", timestamp=" + timestamp +
                ", upload=" + isUpload +
                '}'
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val blackWord = other as BlackWord
        if (stat != blackWord.stat) return false
        if (timestamp != blackWord.timestamp) return false
        if (isUpload != blackWord.isUpload) return false
        if (if (id != null) id != blackWord.id else blackWord.id != null) return false
        return if (word != null) word == blackWord.word else blackWord.word == null
    }

    override fun hashCode(): Int {
        var result = if (id != null) id.hashCode() else 0
        result = 31 * result + if (word != null) word.hashCode() else 0
        result = 31 * result + stat
        result = 31 * result + (timestamp xor (timestamp ushr 32)).toInt()
        result = 31 * result + if (isUpload) 1 else 0
        return result
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        id.apply {
            if (this == null) {
                dest.writeByte(0.toByte())
            } else {
                dest.writeByte(1.toByte())
                dest.writeLong(this)
            }
        }
        dest.writeString(word)
        dest.writeInt(stat)
        dest.writeLong(timestamp)
        dest.writeByte((if (isUpload) 1 else 0).toByte())
    }

    fun mergeFrom(blackWord: BlackWord) {
        word = blackWord.word
        stat = blackWord.stat
        timestamp = blackWord.timestamp
    }

    @IntDef(NORMAL, HIDE, DEL)
    annotation class BlackWordFLag
    companion object {
        const val NORMAL = 0
        const val HIDE = 1
        const val DEL = 2
        private const val TimeFormat = "yyyy-MM-dd HH:mm"
        @JvmField
        val CREATOR: Parcelable.Creator<BlackWord?> = object : Parcelable.Creator<BlackWord?> {
            override fun createFromParcel(`in`: Parcel): BlackWord {
                return BlackWord(`in`)
            }

            override fun newArray(size: Int): Array<BlackWord?> {
                return arrayOfNulls(size)
            }
        }
    }
}
